lib/fast-path.js

```diff
@@ -0,0 +1,202 @@
+  this.stack = [value];
+}
+
+const FPp = FastPath.prototype;
+module.exports = FastPath;
+
+// Static convenience function for coercing a value to a FastPath.
+FastPath.from = function(obj) {
+  return obj instanceof FastPath
+    // Return a defensive copy of any existing FastPath instances.
+    ? obj.copy()
+    // Otherwise use obj as the value of the new FastPath instance.
+    : new FastPath(obj);
+};
+
+FPp.copy = function copy() {
+  const copy = Object.create(FastPath.prototype);
+  copy.stack = this.stack.slice(0);
+  return copy;
+};
+
+// The name of the current property is always the penultimate element of
+// this.stack, and always a String.
+FPp.getName = function getName() {
+  const s = this.stack;
+  const len = s.length;
+  if (len > 1) {
+    return s[len - 2];
+  }
+  // Since the name is always a string, null is a safe sentinel value to
+  // return if we do not know the name of the (root) value.
+  return null;
+};
+
+// The value of the current property is always the final element of
+// this.stack.
+FPp.getValue = function getValue() {
+  const s = this.stack;
+  return s[s.length - 1];
+};
+
+FPp.valueIsDuplicate = function () {
+  const s = this.stack;
+  const valueIndex = s.length - 1;
+  return s.lastIndexOf(s[valueIndex], valueIndex - 1) >= 0;
+};
+
+function getNodeHelper(path, count) {
+  const s = path.stack;
+
+  for (let i = s.length - 1; i >= 0; i -= 2) {
+    const value = s[i];
+    if (isNodeLike(value) && --count < 0) {
+      return value;
+    }
+  }
+
+  return null;
+}
+
+// Without a complete list of Node .type names, we have to settle for this
+// fuzzy matching of object shapes. However, the infeasibility of
+// maintaining a complete list of type names is one of the reasons we're
+// using the FastPath/Visitor abstraction in the first place.
+function isNodeLike(value) {
+  return value &&
+    typeof value === "object" &&
+    ! Array.isArray(value) &&
+    isCapitalized(value.type) &&
+    typeof value.start === "number" &&
+    typeof value.end === "number";
+}
+
+const codeOfA = "A".charCodeAt(0);
+const codeOfZ = "Z".charCodeAt(0);
+function isCapitalized(string) {
+  if (typeof string !== "string") return false;
+  const code = string.charCodeAt(0);
+  return code >= codeOfA && code <= codeOfZ;
+}
+
+FPp.getNode = function getNode(count) {
+  return getNodeHelper(this, ~~count);
+};
+
+FPp.getParentNode = function getParentNode(count) {
+  return getNodeHelper(this, ~~count + 1);
+};
+
+// The length of the stack can be either even or odd, depending on whether
+// or not we have a name for the root value. The difference between the
+// index of the root value and the index of the final value is always
+// even, though, which allows us to return the root value in constant time
+// (i.e. without iterating backwards through the stack).
+FPp.getRootValue = function getRootValue() {
+  const s = this.stack;
+  if (s.length % 2 === 0) {
+    return s[1];
+  }
+  return s[0];
```

Comment:
Even cuter: `return s[(s.length + 1) % 2]`

Though I appreciate the feedback, I think I'm actually going to remove this method from the file for now, since it isn't used yet.